feat: .2dja support for poster.lua

SammyForReal opened the pull request:
This pull request adds support for .2dja files, a file format used by the community to simplify the printing of large signs.

While maintaining compatibility, two optional numbers can now be appended to the syntax to indicate the interval of pages to be used if a .2dja file is provided. Or a single specific page.

Additionally, the status text has been updated.
This commit also ensures that the program is now compatible with normal PCs that don't support all colors.

Finally, comments and style changes have been included for improved readability.

I tried to ensure that the program was as robust and secure as possible.

SammyForReal committed:
feat: .2dja support for poster.lua

Add support for .2dja files, a file format used by the community to simplify the printing of large signs, which essentially contains multiple .2dj files.

Additionally, the status text has been updated.

While maintaining compatibility, two optional numbers can now be appended to the syntax to indicate the interval of pages to be used if a .2dja file is provided. Or a single specific page.

This commit also ensures that the program is now compatible with normal PCs that don't support all colors.

Finally, comments and style changes have been included for improved readability.

emmachase commented:
Thanks for the PR!

A few notes:

Why does this PR include rewriting variable names to use Hungarian notation? This is not consistent with our other ROM programs. (Also FYI, CC-Tweaked ROM is [migrating away from it in general](https://github.com/cc-tweaked/CC-Tweaked/pull/1273#discussion_r1058893385))

Also, for these changes:
> Additionally, the status text has been updated.
> This commit also ensures that the program is now compatible with normal PCs that don't support all colors.

We should make the same changes in the print3d program for consistency.

emmachase commented on src/main/resources/data/computercraft/lua/rom/programs/poster.lua:
```diff
@@ -98,85 +184,90 @@ local function commitPrint(data)
+term.setCursorPos(1, nY+3)
+if bTerminated then
+  term.setTextColor(nErrColor)
+  print(("Printed %s posters out of %s."):format(nPrinted+1, nMaxPages))
+else
+  term.setTextColor(nGreenColor)
+  print(("Printed %s posters!"):format(nPrinted))
```
Not keen on losing proper grammar. This will output "Printed 1 posters!", and "Printed 1 posters out of 3."

emmachase replied:
Also shouldn't these be using %d instead of %s?

emmachase commented on src/main/resources/data/computercraft/lua/rom/programs/poster.lua:
```diff
@@ -98,85 +184,90 @@ local function commitPrint(data)
-while true do
-  local e, p1 = os.pullEventRaw()
-  local n = math.min(math.max((count - remaining) + 1, 1), count)
+for i,data in ipairs(tPosters) do
```
nit: there should be a space after commas

emmachase commented on src/main/resources/data/computercraft/lua/rom/programs/poster.lua:
```diff
@@ -98,85 +184,90 @@ local function commitPrint(data)
+      end
 
+      term.setCursorPos(1,nY+1)
+      printRow("Status: ", (" %s%% : %s / %s "):format(progress, nPrinted, nMaxPages), (" (x%s)       "):format(nCount))
```
Wrong formatting codes.

emmachase commented on src/main/resources/data/computercraft/lua/rom/programs/poster.lua:
```diff
@@ -98,85 +184,90 @@ local function commitPrint(data)
-      print("Printed " .. n .. " item" .. (n ~= 1 and "s" or ""))
+  local cur,max = printer.getInkLevel()
+  term.setCursorPos(1,nY+2)
+  printRow("Ink: ", (" %s%%   "):format(math.floor(100/max*cur)))
```
Wrong formatting codes.

emmachase commented on src/main/resources/data/computercraft/lua/rom/programs/poster.lua:
```diff
@@ -98,85 +184,90 @@ local function commitPrint(data)
+  printRow("Name: ", (data.label or "???"):sub(1,nW-6))
+  
+  term.setCursorPos(1,nY+1)
+  printRow("Status: ", (" %s%% : %s / %s "):format(0, nPrinted, nMaxPages), (" (x%s)       "):format(nCount))
```
Wrong formatting codes.

emmachase commented on src/main/resources/data/computercraft/lua/rom/programs/poster.lua:
```diff
@@ -1,94 +1,180 @@
+    print("Loading...")
+    sName = shell.resolve(sName)
+    if not fs.exists(sName) then
+      -- Try to find a .2dj or .2dja file if the extension did not got specified 
+      if not sFilename:match(".2dj$") or not sFilename:match(".2dja$") then
+        for _,sEnd in ipairs({".2dj", ".2dja"}) do
+          local newName = sName..sEnd
+          if fs.exists(newName) then
+            return loadPoster(newName)
+          end
         end
       end
```
If both exist, we should not assume one or the other, we should error and force the user to be more specific.

emmachase commented on src/main/resources/data/computercraft/lua/rom/programs/poster.lua:
```diff
@@ -1,94 +1,180 @@
 local printer = peripheral.find("poster_printer")
 if not printer then
-  error("No printer found", 0)
+  error("No printer found.", 0)
```
This is inconsistent with CraftOS error patterns: https://github.com/search?q=repo%3Acc-tweaked%2FCC-Tweaked+error%28%22+language%3ALua&type=code&l=Lua

SammyForReal commented:
Could you maybe explain what exactly you mean with `Wrong formatting codes`? Is it "only" a reference to wrong comma placing?

SammyForReal committed:
fix: Removed filename completion and added Index out of bounds check

Removed code that searched for similar-looking files and replaced it with an error message.

Also added an error check for when the given bounds exceed the maximum number of pages in the table.

Additionally, made some minor code style changes including:
* Removing Hungarian notation
* Using %d for numbers instead of %s
* Standardizing comma spacing to one space
* More consistent error patterns
* Ensuring proper grammar.

SammyForReal commented:
I applied all requested changes. However, I left the ones open where I'm not sure if I have fully fulfilled them, which is why I request small feedback please.

ActuallyRuben commented:
> Could you maybe explain what exactly you mean with Wrong formatting codes? Is it "only" a reference to wrong comma placing?

`string.format` uses the same format string rules as [printf](https://linux.die.net/man/3/printf) in C. The short version is that you should use `%s` for string parameters, `%d` for integers and `%f` for decimal values.

It also allows you more control over the way your numbers are presented, as an example using `%5.1f` would format a number as 5 characters long with one number after the decimal point (3 digits, a dot then another digit makes 5 characters).

The following snippet could show the following percentages as an example
```lua
printRow("Ink:", ("%5.f%%   "):format(math.floor(100/max*cur)))
```
```
Ink: 100.0%
Ink:  52.5%
Ink:  12.3%
Ink:   0.0%
```
Look at how nice they align

SammyForReal commented:
Ah I see! Thanks for taking the time explaining me this. I will overview this again.

SammyForReal committed:
fix: Status text wont exit after first print and chore: Text alignement

Because I renamed all variables to be more consistent, there were some redeclarations that accured the program in exiting after the first print.

In addition, formating got corrected to  use %s, %d or %f more properly, including spacing.

SammyForReal committed:
chore: Text Alignment and number formating

Text Alignment for the statuses is more space efficient.

Instead of using hardcoded sizes for height, the actual height is used now.

Because the percentages are perfectly rounded anyways (even without math.floor), %d is used instead.

MineRobber9000 commented on src/main/resources/data/computercraft/lua/rom/programs/poster.lua:
```diff
@@ -27,68 +37,137 @@ if args[1] == "stop" then
+    print("Loading...")
     name = shell.resolve(name)
     if not fs.exists(name) then
-      -- Try to find a .2dj file if the user didn't specify the extension
```
This should either be kept or configured as a `poster.default_extension` setting IMO. No point in removing it.

emmachase replied:
I don't think it should be a setting, but I agree we should not remove this behavior. Sorry if I was unclear in my previous comment. The behavior should be as follows:

1. Attempt to resolve the name as is
2. If that fails, attempt to resolve **both** `.2dj` and `.2dja`
  a. If both exist, error and require the user to specify
  b. If only `.2dj` (no `.2dja`) exists, use that
  c. If only `.2dja` (no `.2dj`) exists, use that
  d. Obviously if neither exist, error

SammyForReal replied:
Ah, yeah that makes more sense. My bad.

emmachase commented on src/main/resources/data/computercraft/lua/rom/programs/poster.lua:
```diff
@@ -27,68 +37,137 @@ if args[1] == "stop" then
+    index.to = index.from
+  end
+  
+  if index.from then
+    index.all = false
+  end
```
This condition is always true.

emmachase commented on src/main/resources/data/computercraft/lua/rom/programs/poster.lua:
```diff
@@ -27,68 +37,137 @@ if args[1] == "stop" then
+local index = { all = true, from = nil, to = nil }
+if #args >= 3 then
+  local num1, num2 = getNum(args[3]), getNum(args[4])
+  if not num1 or not num2 then
+    printUsage()
+    return
+  end
```
Have you tested this works the way you wanted?

Unless I'm reading this wrong, it looks like this will printUsage if args[4] is nil (but args[3] is valid), in which case specifying only an index would not work.

SammyForReal replied:
The idea behind this code is that if either of the given `args` is not a valid number, the usage will be printed, as something else that we can not accept got provided. `getNum` does the check for us already.

If nothing was provided in `args[3]` and `args[4]`, we expect that the user wants the whole file to be printed.

emmachase replied:
If only args[3] is provided, e.g. `poster test.2dja 1 8`, based on your help text this should print only the 8th page. But based on this code, it will error out since args[4] is nil.

SammyForReal replied:
Oh I see. Yeah I'll fix that.

emmachase commented on src/main/resources/data/computercraft/lua/rom/programs/poster.lua:
```diff
@@ -27,68 +37,137 @@ if args[1] == "stop" then
 end
 
-print("")
+-- Setup colors
+local highlightColor = colors.lightGray
+local errColor = colors.gray
+local greenColor = colors.white
+
+if term.isColor() then
+  highlightColor = colors.yellow
+  errColor = colors.red
+  greenColor = colors.green
+end
```
I think this is the wrong place to do this, I would prefer the colors be calculated at the top of the file (right next to `local startTColor = term.getTextColor()`)

emmachase commented on src/main/resources/data/computercraft/lua/rom/programs/poster.lua:
```diff
@@ -101,82 +180,87 @@ local function commitPrint(data)
+  commitPrint(data)
+  
+  term.setCursorPos(1, ypos)
+  printRow("Name: ", (data.label or "???"):sub(1, width-6))
```
This should default to something more friendly, perhaps `Untitled` or `None`

emmachase commented on src/main/resources/data/computercraft/lua/rom/programs/poster.lua:
```diff
@@ -101,82 +180,87 @@ local function commitPrint(data)
+term.setCursorPos(1, ypos+3)
+if terminated then
+  term.setTextColor(errColor)
+  print(("Printed %d poster out of %d."):format(printed+1, maxPages))
+else
+  term.setTextColor(greenColor)
+  print(("Printed %d poster!"):format(printed))
```
This is still incorrect. It should be `poster` if equal to 1, `posters` otherwise.

SammyForReal committed:
fix: poster.lua syntax for intervals and filename completion

The following bugs, mostly realted with the syntax got fixed:

* If a valid third argument is given without a fourth, it is taken as an index for .2dja files. If a fourth argument is also given, it is interpreted as an interval for it. If no arguments are provided, all pages are assumed to be printed. If an invalid argument is provided, the usage is printed.
* If a filename is provided without an extension, a file with a known extension will be searched for. If multiple files are found, or none are found, an corresponding error message is displayed. If a single file is found, it is used as intended.

Finally, some printed messages were changed too, such as the title and final message indicating the number of posters printed."
